### PR TITLE
[20040] Merge branch 'dev' into feature/rails4

### PR DESCRIPTION
**Pull request by myabc**
https://community.openproject.org/work_packages/20040

This brings in latest changes from `dev` into #2104 - since this was a non-trivial merge, I'm issuing this as a PR.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://myabc-feature-merge-dev-into-rails4.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-merge-dev-into-rails4.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by NobodysNightmare** [comment]
> since this was a non-trivial merge, I'm issuing this as a PR.

There is not list of conflicting files etc., is there?
I am trying my luck with `git show e4c87bc`, which reduces the diff quite a lot, but it still seems to be rather noisy...

If you have no better idea, I will merge this PR after my look at the output of `git show` and green specs... I think this is how this should work right? Specs tell the truth...

**Comment by NobodysNightmare** [comment]
The three failures in core specs look legitimate.

I can give more context to the attachment service spec, if that helps:

When adding an attachment we want to ensure that an error in saving an attachments journal entry (on the WP), will also cause the attachment to not be created. That is: creating attachment + journal should be atomic. The classical way of doing this is using database transactions, which was the approach chosen. I am not sure why this should stop to work with Rails 4. Granted: The spec injects the saving error quite bluntly, but I don't see why this should cause issues for that spec.

**tl; dr:** Sadly I don't know whats wrong here, but maybe the context I've given is enough for you, to come to a conclusion ^^

**Comment by myabc** [comment]
@oliverguenther you might want to run `rubocop -a` on some of the Repository-related specs.

**Comment by ulferts** [comment]
I can't reproduce those failures locally:

```
Failures:

  1) TimelogController should index all projects csv export
     [31mFailure/Error:[0m [31massert response.body.include?("\n04/21/2007,redMine Admin,Design,eCookbook,3,Bug,Error 281 when updating a recipe,1.0,\"\"\n")[0m
     [31mMiniTest::Assertion[0m:
       [31mFailed assertion, no message given.[0m
[36m     # ./spec/legacy/functional/timelog_controller_spec.rb:257:in `block (2 levels) in <top (required)>'[0m

  2) TimelogController should index csv export
     [31mFailure/Error:[0m [31massert response.body.include?("\n04/21/2007,redMine Admin,Design,eCookbook,3,Bug,Error 281 when updating a recipe,1.0,\"\"\n")[0m
     [31mMiniTest::Assertion[0m:
       [31mFailed assertion, no message given.[0m
[36m     # ./spec/legacy/functional/timelog_controller_spec.rb:266:in `block (2 levels) in <top (required)>'[0m

Finished in 3 minutes 0.9 seconds
[31m378 examples, 2 failures[0m

Failed examples:

[31mrspec ./spec/legacy/functional/timelog_controller_spec.rb:251[0m [36m# TimelogController should index all projects csv export[0m
[31mrspec ./spec/legacy/functional/timelog_controller_spec.rb:260[0m [36m# TimelogController should index csv export[0m
```

**Comment by myabc** [comment]
> I can't reproduce those failures locally:

@ulferts I can't either. I suspected it might have been a TZ issue, although my fix forcing a particular TZ didn't work  - you might be able to get the spec working by removing the `04/21/2007`.

**Comment by ulferts** [comment]
@myabc I didn't want to go digging into the specs. I have therefore issued myabc#4 that rewrites the failing specs in the hope of them going green. As the failing specs are legacy specs, this is a valiant deed in itself.

**Comment by myabc** [comment]
@ulferts hopefully this is now mergeable.
